Update K8s operator doc with new features (#3331)
wso2/docs-apim: en/docs/install-and-setup/setup/kubernetes-operators/k8s-api-operator/manage-integration/integration-deployments.md

When the integration is successfully deployed, it should create the `hello-world` integration, `hello-world-deployment`, `hello-world-service`, and `ei-operator-ingress` as follows:

!!! Tip

>     The `ei-operator-ingress` will not be created if you have [disabled the ingress controller](#Disable-ingress-controller-creation).

Reviewer: ei-operator-ingress?
Do we still use 'ei'?

Author: This needs a code change. Shall we take it in a different PR?

Reviewer: +1

    {"Hello":"World"}%
    ```

!!! Tip

>     The `ei-operator-ingress` will not be created if you have [disabled the ingress controller](#Disable-ingress-controller-creation).

Reviewer: ei-operator-ingress?
Do we still use 'ei'?

Reviewer 2: Here, separate ingress is created for only integration based deployments by the operator. Ingress creation can be customized via ConfigMap. However, I suggest **integration-ingress** instead of `ei-operator-ingress` here. 
@nilminiwso2 @abeykoon WDYT?

Author: As this needs a code change, will address in a different PR.

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: integration-config
data:
  enableAutoScale: "true"
  maxReplicas: "5"
```

Note how you can set required resources and resource limits for the pods in the deployment referring to the above section [Managing resources for pods in EI deployment](#Managing-resources-for-pods-in-EI-deployment). HPA configs are injected through the integration configmap. See how you can define it at [integration_controller_conf.yaml]([integration_controller_conf.yaml](https://github.com/wso2/k8s-api-operator/blob/master/api-operator/deploy/controller-configs/integration_controller_conf.yaml)) file. 

```yaml
  hpaMetrics: |
    - type: Resource
      resource:
        name: cpu
        target:
          type: Utilization
          averageUtilization: 50
```

## Liveness and readiness probes 

The operator gives you the flexibility to define liveness and readiness probes. These probes are used by Kubernetes to know whether pods are live and whether they are ready to accept traffic. 

> You can define the default probe definitions at `integration_controller_conf.yaml` file and include them to the configMap. Usually, WSO2 micro integrator listens for HTTP traffic on 8290 port. Liveness of the pod is checked by a ping to that port. WSO2 micro integrator is ready to accept traffic when all the CApps are successfully deployed. The API with path `/healthz` which is exposed through port 9201 is used for that. Note that these ports can change as per the configurations used at the MI based image. 

Reviewer 2: Shall we change this phase to "listens for HTTP traffic on passthrough port(default 8290)"?

Reviewer 2: Also, "HTTP inbound port(default 9201)" instead 9201

Author: +1. Updated.

    ```bash
    curl --cacert <CERT_FILE> https://<HOST-NAME>/<INTEGRATION-NAME>-inbound/<PORT>/<CONTEXT>
    ```    

> ## Managing resources for pods in EI deployment

Reviewer: ```suggestion
## Manage resources for pods in a deployment
```

Reviewer 2: I suggest **Integration** instead of EI?

Author: updated to `Manage resources for pods of integration deployment`

> When creating a deployment operator allows to define required resources to run pods created by that deployment. In the same manner, you can define resource limits as well. Following example shows how to set CPU and memory resources at `integration_cr.yaml` file.

Reviewer: ```suggestion
The K8s operator allows you to define the resources that are required for running the pods in a deployment. You can also define resource limits when you define these resources. The following example shows how CPU and memory resources are configured in the `integration_cr.yaml` file.
```

Author: +1. Updated. Thanks!

```yaml
apiVersion: wso2.com/v1alpha2
kind: Integration
metadata:
  name: test-integration
spec:
  image: <MI based Docker image for integration>
  deploySpec:
    minReplicas: 1
    requestCPU: "500m"
    reqMemory: "512Mi"
    cpuLimit: "2000m"
    memoryLimit: "2048Mi"
```

> You can set the default resource configs at integration_controller_conf.yaml and add them to the configMap. Whenever they are not defined at `integration_cr.yaml` file, operator will pick defaults from the configMap. 

Reviewer: ```suggestion
If you don't have any resources defined in the `integration_cr.yaml` file, the K8s operator refers the default resource configurations in the `integration_controller_conf.yaml` file. Therefore, be sure to update the `configMap` section in the `integration_controller_conf.yaml` with default resource configurations. See the example given below.
```

Author: Updated.

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: integration-config
data:
  requestCPU: "500m"
  reqMemory: "512Mi"
  cpuLimit: "2000m"
  memoryLimit: "2048Mi"
```

> Please refer to [Managing Resources for Containers](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/) to learn more. 

Reviewer: ```suggestion
See [Managing Resources for Containers](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/) in the Kubernetes documentation for more details. 
```

Author: Updated.

## Enable Auto Scaling for deployment

> When traffic is increased your pods may need to horizontally scaled. Kubernetes provides a way to define resource limits and a policy for the deployment so that it will auto scale the deployment looking at resource usage. Please read on [Horizontal Pod Scaling (HPA)](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/) in Kubernetes.

Reviewer: ```suggestion
When the traffic to your pods increase, the deployment may need to scale horizontally. 
Kubernetes allows you to define the resource limits and policy in a way that the deployment can auto scale based on resource usage. 

See [Horizontal Pod Scaling (HPA)](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/) in the Kubernetes documentation for details.
```

Author: Updated. Thanks!

> You can enable scaling and define maximum number of replicas to scale under `autoScale` of `integration_cr.yaml` as below. 

Reviewer: ```suggestion
You can enable auto scaling and define the maximum number of replicas to scale by updating the `autoScale` section in the `integration_cr.yaml` file as shown below. 
```

Author: Updated.

  ```yaml
apiVersion: wso2.com/v1alpha2
kind: Integration
metadata:
  name: test-integration
spec:
  autoScale:
    enabled: "true"
    maxReplicas: 3
  ```

> You can enable auto scaling by default by defining it at `integration_controller_conf.yaml` file. These values will get overridden by what is defined at  `integration_cr.yaml` file. 

Reviewer: ```suggestion
If you don't have auto scaling defined in the `integration_cr.yaml` file, the K8s operator refers the default configurations in the `integration_controller_conf.yaml` file. Therefore, be sure to update the `integration_controller_conf.yaml` file with auto scaling configurations as shown below. 
```

Author: +1. Updated.

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: integration-config
data:
  enableAutoScale: "true"
  maxReplicas: "5"
```

Note how you can set required resources and resource limits for the pods in the deployment referring to the above section [Managing resources for pods in EI deployment](#Managing-resources-for-pods-in-EI-deployment). HPA configs are injected through the integration configmap. See how you can define it at [integration_controller_conf.yaml]([integration_controller_conf.yaml](https://github.com/wso2/k8s-api-operator/blob/master/api-operator/deploy/controller-configs/integration_controller_conf.yaml)) file. 

```yaml
  hpaMetrics: |
    - type: Resource
      resource:
        name: cpu
        target:
          type: Utilization

>           averageUtilization: 50

Reviewer: When the CPU utilisation for one pod exceeds this value, the scaling starts? Is that how it works?

@abeykoon

Author: There is an algorithm. Quoting K8s docs, 

```
For per-pod resource metrics (like CPU), the controller fetches the metrics from the resource metrics API for each Pod targeted by the HorizontalPodAutoscaler. Then, if a target utilization value is set, the controller calculates the utilization value as a percentage of the equivalent resource request on the containers in each Pod. If a target raw value is set, the raw metric values are used directly. The controller then takes the mean of the utilization or the raw value (depending on the type of target specified) across all targeted Pods, and produces a ratio used to scale the number of desired replicas.
```

```

## Liveness and readiness probes 

> The operator gives you the flexibility to define liveness and readiness probes. These probes are used by Kubernetes to know whether pods are live and whether they are ready to accept traffic. 

Reviewer: ```suggestion
The operator gives you the flexibility of defining liveness and readiness probes. These probes are used by Kubernetes to know whether the pods are live and whether they are ready to accept traffic. 
```

Author: Updated.

> You can define the default probe definitions at `integration_controller_conf.yaml` file and include them to the configMap. Usually, WSO2 micro integrator listens for HTTP traffic on passthrough port(default 8290). Liveness of the pod is checked by a ping to that port. WSO2 micro integrator is ready to accept traffic when all the CApps are successfully deployed. The API with path `/healthz` which is exposed through HTTP inbound port(default 9201) is used for that. Note that these ports can change as per the configurations used at the MI based image. 

Reviewer: ```suggestion
You can configure the default probe definition in the `integration_controller_conf.yaml` file as shown below. The Note that the Micro Integrator is considered **live** and **ready** to accept traffic only when all the CApps are successfully deployed.

Usually, the Micro Integrator listens for HTTP traffic on the passthrough port (default 8290). The liveness of the pod is checked by a ping to that port. The API with the `/healthz` path, which is exposed through the HTTP inbound port (default 9201), is used for this purpose. These ports can change according to the configurations in the Micro Integrator based image. 
```

Author: The suggested changes the meaning.

Author: Updated in a way that there is no harm to the meaning. @nilminiwso2 please check.

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: integration-config
data:
  livenessProbe: |
    tcpSocket:
      port: 8290
    initialDelaySeconds: 10
    periodSeconds: 5
  readinessProbe: |
    httpGet:
      path: /healthz
      port: 9201
    initialDelaySeconds: 10
    periodSeconds: 5
```

> Use `integration_cr.yaml` file to define the probes specific to a particular deployment like below. 

Reviewer: ```suggestion
Use the `integration_cr.yaml` file to define the probes specific to a particular deployment. See the example given below. 
```

```yaml
apiVersion: wso2.com/v1alpha2
kind: Integration
metadata:
  name: test-integration
spec:
  image: <MI based Docker image for integration>
  deploySpec:
    livenessProbe:
      tcpSocket:
        port: 8290
      initialDelaySeconds: 30
      periodSeconds: 10
    readinessProbe:
      httpGet:
        path: /healthz
        port: 9201
      initialDelaySeconds: 30
      periodSeconds: 10
```

Note that any you can configure any configuration supported under these probes as defined at [Kubernetes documentation](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/).

## Update the existing deployment 

> You can update deployment definition at `integration_cr.yaml` file and re-apply it. Then operator will update the existing deployment to match with the new deployment definition. Docker image, exposed ports, resource definitions, environment variables etc can be updated.  

Reviewer: ```suggestion
You can update the configurations in the `integration_cr.yaml` file and reapply it to your existing deployment when you use the K8s operator. The Docker image, exposed ports, resource definitions, environment variables, etc. can be updated.  
```

Author: Updated.

## Additional operator configurations 

### Disable ingress controller creation

> By default, an ingress controller named `ei-operator-ingress` is created to expose all the deployments created by the operator. Per deployment, new rules are added to the same ingress controller to route the traffic. Sometimes you may use an external ingress or define an ingress yourself. In such cases, you can disable ingress controller creation as below. 

Reviewer: ```suggestion
By default, an ingress controller named `ei-operator-ingress` is created to expose all the deployments created by the operator. Per deployment, new rules are added to the same ingress controller to route the traffic. Sometimes you may use an external ingress or define an ingress yourself. In such cases, you can disable ingress controller creation from the `integration_controller_conf.yaml` file as shown below. 
```

Author: Updated.

> defined at `integration_controller_conf.yaml` file

Reviewer: ```suggestion
```

Author: Removed.

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: integration-config
data:
  reconcileInterval: "10"
  autoIngressCreation: "false"
```

### Change reconcile interval

> The operator continuously run a task listening for changes applied to the deployment. When you update `integration_cr.yaml` and apply, this is the task that updates the deployment. You can configure how often this task should run at the operator level (in seconds). 

Reviewer: ```suggestion
The K8s operator continuously runs a task that listens for changes applied to the deployment. When you change the `integration_cr.yaml` file and reapply, this is the task that updates the deployment. You can configure how often this task should run at the operator level (in seconds) by using the `integration_controller_conf.yaml` file.
```

Author: Updated.

## Enable Auto Scaling for deployment

When the traffic to your pods increase, the deployment may need to scale horizontally. 
Kubernetes allows you to define the resource limits and policy in a way that the deployment can auto scale based on resource usage. 
See [Horizontal Pod Scaling (HPA)](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/) in the Kubernetes documentation for details.

You can enable auto scaling and define the maximum number of replicas to scale by updating the `autoScale` section in the `integration_cr.yaml` file as shown below. 

  ```yaml
apiVersion: wso2.com/v1alpha2
kind: Integration
metadata:
  name: test-integration
spec:
  autoScale:
    enabled: "true"
    maxReplicas: 3
  ```

If you don't have auto scaling defined in the `integration_cr.yaml` file, the K8s operator refers the default configurations in the `integration_controller_conf.yaml` file. Therefore, be sure to update the `integration_controller_conf.yaml` file with auto scaling configurations as shown below.

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: integration-config
data:
  enableAutoScale: "true"
  maxReplicas: "5"
```

> Note how you can set required resources and resource limits for the pods in the deployment referring to the above section [Managing resources for pods in EI deployment](#Managing-resources-for-pods-in-EI-deployment). HPA configs are injected through the integration configmap. See how you can define it at [integration_controller_conf.yaml]([integration_controller_conf.yaml](https://github.com/wso2/k8s-api-operator/blob/master/api-operator/deploy/controller-configs/integration_controller_conf.yaml)) file. 

Reviewer: ```suggestion
**Resource conditions for pod auto scaling**

The required resources and resource limits for pods in a *single deployment* are defined in the `integration_cr.yaml` file as explained in the [Managing resources for pods in an integration deployment](#Managing-resources-for-pods-in-EI-deployment) section.  However, the resource conditions for auto scaling should be defined as an `hpaMetrics` configuration in the [integration_controller_conf.yaml]([integration_controller_conf.yaml](https://github.com/wso2/k8s-api-operator/blob/master/api-operator/deploy/controller-configs/integration_controller_conf.yaml)) file as shown below. Note that configurations in the `integration_controller_conf.yaml` file are global and not specific to pods in a single deployment.

!!! Info
     See the [Kubernetes documentation](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/) for more details on horizontal pod autoscaling.
```

Author: +1 for the new wording and details.

```yaml
  hpaMetrics: |
    - type: Resource
      resource:
        name: cpu
        target:
          type: Utilization
          averageUtilization: 50
```

## Liveness and readiness probes 

The operator gives you the flexibility of defining liveness and readiness probes. These probes are used by Kubernetes to know whether the pods are live and whether they are ready to accept traffic.

You can configure the default probe definition in the `integration_controller_conf.yaml` file as shown below. 

> Micro Integrator is considered **live** when it is accepting HTTP/HTTPS traffic. Usually, it listens for HTTP traffic on passthrough port(default 8290). Liveness of the MI pod is checked by a ping to that port.

Reviewer: ```suggestion
The Micro Integrator is considered **live** when it is accepting HTTP/HTTPS traffic. Usually, it listens for HTTP traffic on the passthrough port (default 8290). Liveness of the Micro Integrator pod is checked by a ping to that port.
```

> WSO2  Micro Integrator is **ready** to accept traffic only when all the CApps are successfully deployed. The API with path `/healthz` which is exposed through HTTP inbound port(default 9201) is used to check that.

Reviewer: ```suggestion
The  Micro Integrator is **ready** to accept traffic only when all the CApps are successfully deployed. The API with the `/healthz` path, which is exposed through the HTTP inbound port (default 9201) is used to check the readiness.
```

> Note that these ports can change as per the configurations used at the MI based image. 

Reviewer: ```suggestion
!!! Note 
     These ports can change as per the configurations used in the Micro Integrator based image. 
```
